In [1]:
%cd "Data"

C:\Users\khaih\Downloads\LDS0\Data


In [2]:
import nltk

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')
nltk.download('omw-1.4')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\khaih\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\khaih\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\khaih\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\khaih\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\khaih\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


True

In [3]:
import pandas as pd
import numpy as np
from underthesea import word_tokenize, pos_tag, sent_tokenize
import regex
import demoji
from pyvi import ViPosTagger, ViTokenizer
import string

In [4]:
import pandas as pd
from nltk.tokenize import word_tokenize
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.linear_model import LogisticRegression
from sklearn.tree import ExtraTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

from sklearn.metrics import confusion_matrix, classification_report
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from imblearn.over_sampling import SMOTE
from imblearn.over_sampling import RandomOverSampler

In [5]:
import pandas as pd
data = pd.read_csv("pre_data_8.csv", delimiter=',',index_col=0)

In [6]:
data.head(10)

class                                        pre_comment
259179  Like                                            kiểm đồ
259180  Like  cho_phép sơ_suất troâm nhạc hát hoạt_động nhậv...
259181  Like  vừa_vặn nâu trăng dày hơn đỏ trắng đỏ cổ hơn n...
259182  Like                                đầu_shop hợp mùa hè
259183  Like                                đẹp đẹp nhanh thích
259184  Like                                  đẹp dày_dặn nhanh
259185  Like                   mét sáu_mươi_lăm lô gờ fit luônn
259186  Like                     không_không không_nhỉ quên học
259187  Like                                    chê tốt_đẹp hợp
259188  Like         gọn_gàng đẹp ôm tôn dáng hình_ảnh kâng họa

In [7]:
data.isna().sum()

class          0
pre_comment    0
dtype: int64

In [8]:
data.count()

class          526579
pre_comment    526579
dtype: int64

In [9]:
X=data["pre_comment"]
y=data["class"]

In [10]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2,random_state=42)

In [11]:
ros = RandomOverSampler()
X_train_resampled, y_train_resampled = ros.fit_resample(np.array(X_train).reshape(-1, 1), np.array(y_train).reshape(-1, 1));
train_df_reampled = pd.DataFrame(list(zip([x[0] for x in X_train_resampled], y_train_resampled)), columns = ['pre_comment', 'class']);

In [12]:
train_df_reampled["class"].value_counts()

Like        359326
Neutral     359326
Not_Like    359326
Name: class, dtype: int64

In [13]:
train_df_reampled.to_csv("data_resample_1.csv")

In [14]:
X_train_resampled, X_test, y_train_resampled, y_test = train_test_split(train_df_reampled["pre_comment"],train_df_reampled["class"], test_size=0.2,random_state=42)

In [ ]:
train_df_reampled.loc[train_df_reampled["pre_comment"]==""]

In [ ]:
X_train_resampled=train_df_reampled["pre_comment"]
y_train_resampled=train_df_reampled["class"]

In [ ]:
# from sklearn.feature_extraction.text import CountVectorizer

# vectorizer = CountVectorizer()
# vectorizer.fit(X_train.values)
# X_train=vectorizer.transform(X_train.values)
# X_test=vectorizer.transform(X_test.values)
# X_train=X_train.toarray()
# X_test=X_test.toarray()

# x_train = pd.DataFrame(X_train)
# X_train_resampled, y_train_resampled = SMOTE().fit_resample(X_train, y_train)

In [ ]:
# smote = SMOTE(random_state=42)
# X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

In [15]:
def my_tokenizer(text):
    return word_tokenize(text)

### BernoulliNB

In [16]:
pipeline_bnl = Pipeline([
    ('vectorizer', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('model', BernoulliNB())
])

pipeline_bnl.fit(X_train_resampled, y_train_resampled)

y_pred_bnl = pipeline_bnl.predict(X_test)

accuracy_bnl = accuracy_score(y_test, y_pred_bnl)
print("Accuracy: ", accuracy_bnl)
print("-"*60)
print(confusion_matrix(y_test, y_pred_bnl))
print("-"*60)
print(classification_report(y_test, y_pred_bnl))

Accuracy:  0.7905898068609807
------------------------------------------------------------
[[59496 10267  2446]
 [ 4989 55817 10859]
 [ 1876 14711 55135]]
------------------------------------------------------------
              precision    recall  f1-score   support

        Like       0.90      0.82      0.86     72209
     Neutral       0.69      0.78      0.73     71665
    Not_Like       0.81      0.77      0.79     71722

    accuracy                           0.79    215596
   macro avg       0.80      0.79      0.79    215596
weighted avg       0.80      0.79      0.79    215596



### LogisticRegression

In [17]:
pipeline_lr = Pipeline([
    ('vectorizer', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('model', LogisticRegression())
])

pipeline_lr.fit(X_train_resampled, y_train_resampled)

y_pred_lr = pipeline_lr.predict(X_test)

accuracy_lr = accuracy_score(y_test, y_pred_lr)
print("Accuracy:", accuracy_lr)
print("-"*60)
print(confusion_matrix(y_test, y_pred_lr))
print("-"*60)
print(classification_report(y_test, y_pred_lr))

C:\Users\khaih\AppData\Roaming\Python\Python39\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Accuracy: 0.8113972429915212
------------------------------------------------------------
[[62498  7269  2442]
 [ 5238 54457 11970]
 [ 2259 11484 57979]]
------------------------------------------------------------
              precision    recall  f1-score   support

        Like       0.89      0.87      0.88     72209
     Neutral       0.74      0.76      0.75     71665
    Not_Like       0.80      0.81      0.80     71722

    accuracy                           0.81    215596
   macro avg       0.81      0.81      0.81    215596
weighted avg       0.81      0.81      0.81    215596



In [ ]:
from joblib import dump
dump(pipeline_lr, 'Project_1_LogisticRegression_sklearn_3.joblib')

### DecisionTree

In [18]:
pipeline_dct = Pipeline([
    ('vectorizer', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('model', DecisionTreeClassifier())
])

pipeline_dct.fit(X_train_resampled, y_train_resampled)

y_pred_dct = pipeline_dct.predict(X_test)

accuracy_dct = accuracy_score(y_test, y_pred_dct)
print("Accuracy:", accuracy_dct)
print("-"*60)
print(confusion_matrix(y_test, y_pred_dct))
print("-"*60)
print(classification_report(y_test, y_pred_dct))

Accuracy: 0.9260700569583852
------------------------------------------------------------
[[63947  4440  3822]
 [ 1227 67362  3076]
 [  523  2851 68348]]
------------------------------------------------------------
              precision    recall  f1-score   support

        Like       0.97      0.89      0.93     72209
     Neutral       0.90      0.94      0.92     71665
    Not_Like       0.91      0.95      0.93     71722

    accuracy                           0.93    215596
   macro avg       0.93      0.93      0.93    215596
weighted avg       0.93      0.93      0.93    215596



In [20]:
from joblib import dump
dump(pipeline_dct, 'Project_1_DecisionTree_sklearn_4.joblib')

['Project_1_DecisionTree_sklearn_4.joblib']

### ExtraTreeClassifier

In [19]:
pipeline_etc = Pipeline([
    ('vectorizer', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('model', ExtraTreeClassifier())
])

pipeline_etc.fit(X_train_resampled, y_train_resampled)

y_pred_etc = pipeline_etc.predict(X_test)

accuracy_etc = accuracy_score(y_test, y_pred_etc)
print("Accuracy: ", accuracy_etc)
print("-"*60)
print(confusion_matrix(y_test, y_pred_etc))
print("-"*60)
print(classification_report(y_test, y_pred_etc))

Accuracy:  0.9313298948032431
------------------------------------------------------------
[[65081  3387  3741]
 [ 1227 67362  3076]
 [  525  2849 68348]]
------------------------------------------------------------
              precision    recall  f1-score   support

        Like       0.97      0.90      0.94     72209
     Neutral       0.92      0.94      0.93     71665
    Not_Like       0.91      0.95      0.93     71722

    accuracy                           0.93    215596
   macro avg       0.93      0.93      0.93    215596
weighted avg       0.93      0.93      0.93    215596



In [21]:
from joblib import dump
dump(pipeline_dct, 'Project_1_Extra_sklearn_4.joblib')

['Project_1_Extra_sklearn_4.joblib']

In [ ]:
text=["dở_tệ không tiền"]

In [ ]:
text=["dở_tệ không tiền"]
text=pd.DataFrame({'pre_comment':text})
import joblib
loaded_model = joblib.load("Project_1_sklearn.joblib")
result = loaded_model.predict(text["pre_comment"])
result

### Nhận xét:
- Sau khi resample data nhận thấy rằng accuracy giảm đi, nhưng recall và f1-score tăng đáng kể, cho thấy tính hiệu quả được cải thiện.

=> Có thể chấp nhận các model đã chạy qua cách cải thiện resample data. 

=> Model DecisionTree và ExtraTreeClassifier cho ra kết quả tốt nhất và có thể áp dụng nó để predict (Accuracy lần lượt là: 89,6% và 90%), đồng thời các điểm recall, f1-score đều khá cao dao động từ (88%-93%) 